// @ngxs/hmr-plugin/esm5/src/internal/hmr-state-context-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Store } from '@ngxs/store';
import { isStateOperator } from '@ngxs/store/operators';
/**
 * @template T, S
 */
var /**
 * @template T, S
 */
HmrStateContextFactory = /** @class */ (function () {
    function HmrStateContextFactory(module) {
        /** @type {?} */
        var store = module.injector.get(Store, undefined);
        if (!store) {
            throw new Error('Store not found, maybe you forgot to import the NgxsModule');
        }
        this.store = store;
    }
    /**
     * @description
     * must be taken out into  @ngxs/store/internals
     */
    /**
     * \@description
     * must be taken out into  \@ngxs/store/internals
     * @return {?}
     */
    HmrStateContextFactory.prototype.createStateContext = /**
     * \@description
     * must be taken out into  \@ngxs/store/internals
     * @return {?}
     */
    function () {
        var _this = this;
        return {
            dispatch: (/**
             * @param {?} actions
             * @return {?}
             */
            function (actions) { return (/** @type {?} */ (_this.store)).dispatch(actions); }),
            getState: (/**
             * @return {?}
             */
            function () { return (/** @type {?} */ ((/** @type {?} */ (_this.store)).snapshot())); }),
            setState: (/**
             * @param {?} val
             * @return {?}
             */
            function (val) {
                if (isStateOperator(val)) {
                    /** @type {?} */
                    var currentState = (/** @type {?} */ (_this.store)).snapshot();
                    val = val(currentState);
                }
                (/** @type {?} */ (_this.store)).reset(val);
                return (/** @type {?} */ (val));
            }),
            patchState: (/**
             * @param {?} val
             * @return {?}
             */
            function (val) {
                /** @type {?} */
                var currentState = (/** @type {?} */ (_this.store)).snapshot();
                /** @type {?} */
                var newState = tslib_1.__assign({}, currentState, ((/** @type {?} */ (val))));
                (/** @type {?} */ (_this.store)).reset(newState);
                return newState;
            })
        };
    };
    return HmrStateContextFactory;
}());
/**
 * @template T, S
 */
export { HmrStateContextFactory };
if (false) {
    /** @type {?} */
    HmrStateContextFactory.prototype.store;
}
