// @ngxs/hmr-plugin/esm5/src/internal/hmr-lifecycle.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HmrBeforeDestroyAction } from '../actions/hmr-before-destroy.action';
/**
 * @template T, S
 */
var /**
 * @template T, S
 */
HmrLifecycle = /** @class */ (function () {
    function HmrLifecycle(ngAppModule, bootstrap, storage, context, options) {
        this.ngAppModule = ngAppModule;
        this.bootstrap = bootstrap;
        this.storage = storage;
        this.context = context;
        this.options = options;
    }
    /**
     * @param {?} hmrAfterOnInit
     * @return {?}
     */
    HmrLifecycle.prototype.hmrNgxsStoreOnInit = /**
     * @param {?} hmrAfterOnInit
     * @return {?}
     */
    function (hmrAfterOnInit) {
        /** @type {?} */
        var moduleHmrInit = this.getModuleHmrInitCallback();
        moduleHmrInit = moduleHmrInit.bind(this.ngAppModule);
        this.stateEventLoop((/**
         * @param {?} ctx
         * @param {?} state
         * @return {?}
         */
        function (ctx, state) {
            moduleHmrInit(ctx, state);
            hmrAfterOnInit(ctx, state);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    HmrLifecycle.prototype.getModuleHmrInitCallback = /**
     * @private
     * @return {?}
     */
    function () {
        if (typeof this.ngAppModule.hmrNgxsStoreOnInit === 'function') {
            return this.ngAppModule.hmrNgxsStoreOnInit;
        }
        return (/**
         * @param {?} ctx
         * @param {?} state
         * @return {?}
         */
        function defaultModuleHmrInit(ctx, state) {
            ctx.patchState(state);
        });
    };
    /**
     * @return {?}
     */
    HmrLifecycle.prototype.hmrNgxsStoreBeforeOnDestroy = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var state = {};
        /** @type {?} */
        var ctx = this.context.createStateContext();
        if (typeof this.ngAppModule.hmrNgxsStoreBeforeOnDestroy === 'function') {
            state = this.ngAppModule.hmrNgxsStoreBeforeOnDestroy(ctx);
        }
        else {
            state = ctx.getState();
        }
        ctx.dispatch(new HmrBeforeDestroyAction(state));
        return state;
    };
    /**
     * @private
     * @param {?} callback
     * @return {?}
     */
    HmrLifecycle.prototype.stateEventLoop = /**
     * @private
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        var _this = this;
        if (!this.storage.hasData())
            return;
        /** @type {?} */
        var appBootstrapped$ = this.bootstrap.appBootstrapped$;
        /** @type {?} */
        var state$ = this.context.store.select((/**
         * @param {?} state
         * @return {?}
         */
        function (state) { return state; }));
        appBootstrapped$.subscribe((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var eventId;
            /** @type {?} */
            var storeEventId = state$.subscribe((/**
             * @return {?}
             */
            function () {
                // setTimeout used for zone detection after set hmr state
                clearInterval(eventId);
                eventId = window.setTimeout((/**
                 * @return {?}
                 */
                function () {
                    // close check on the message queue
                    storeEventId.unsubscribe();
                    // if events are no longer running on the call stack,
                    // then we can update the state
                    callback(_this.context.createStateContext(), (/** @type {?} */ (_this.storage.snapshot)));
                }), _this.options.deferTime);
            }));
        }));
    };
    return HmrLifecycle;
}());
/**
 * @template T, S
 */
export { HmrLifecycle };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HmrLifecycle.prototype.ngAppModule;
    /**
     * @type {?}
     * @private
     */
    HmrLifecycle.prototype.bootstrap;
    /**
     * @type {?}
     * @private
     */
    HmrLifecycle.prototype.storage;
    /**
     * @type {?}
     * @private
     */
    HmrLifecycle.prototype.context;
    /**
     * @type {?}
     * @private
     */
    HmrLifecycle.prototype.options;
}
