// angular-gojs/bundles/angular-gojs.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-gojs', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['angular-gojs'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var GojsComponent = (function () {
        function GojsComponent() {
        }
        /**
         * @return {?}
         */
        GojsComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        GojsComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'gojs-view',
                        template: "<p>\n  gojs\n</p>\n",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        GojsComponent.ctorParameters = function () { return []; };
        return GojsComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var GojsViewModule = (function () {
        function GojsViewModule() {
        }
        GojsViewModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [GojsComponent],
                        exports: [
                            GojsComponent
                        ]
                    },] }
        ];
        return GojsViewModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.GojsViewModule = GojsViewModule;
    exports.ɵa = GojsComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
