// @ngxs/hmr-plugin/esm2015/src/internal/hmr-state-context-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Store } from '@ngxs/store';
import { isStateOperator } from '@ngxs/store/operators';
/**
 * @template T, S
 */
export class HmrStateContextFactory {
    /**
     * @param {?} module
     */
    constructor(module) {
        /** @type {?} */
        const store = module.injector.get(Store, undefined);
        if (!store) {
            throw new Error('Store not found, maybe you forgot to import the NgxsModule');
        }
        this.store = store;
    }
    /**
     * \@description
     * must be taken out into  \@ngxs/store/internals
     * @return {?}
     */
    createStateContext() {
        return {
            dispatch: (/**
             * @param {?} actions
             * @return {?}
             */
            actions => (/** @type {?} */ (this.store)).dispatch(actions)),
            getState: (/**
             * @return {?}
             */
            () => (/** @type {?} */ ((/** @type {?} */ (this.store)).snapshot()))),
            setState: (/**
             * @param {?} val
             * @return {?}
             */
            val => {
                if (isStateOperator(val)) {
                    /** @type {?} */
                    const currentState = (/** @type {?} */ (this.store)).snapshot();
                    val = val(currentState);
                }
                (/** @type {?} */ (this.store)).reset(val);
                return (/** @type {?} */ (val));
            }),
            patchState: (/**
             * @param {?} val
             * @return {?}
             */
            val => {
                /** @type {?} */
                const currentState = (/** @type {?} */ (this.store)).snapshot();
                /** @type {?} */
                const newState = Object.assign({}, currentState, ((/** @type {?} */ (val))));
                (/** @type {?} */ (this.store)).reset(newState);
                return newState;
            })
        };
    }
}
if (false) {
    /** @type {?} */
    HmrStateContextFactory.prototype.store;
}
