// @ngxs/hmr-plugin/esm2015/src/symbols.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const HmrRuntime = {
    Status: 'NGXS_HMR_LIFECYCLE_STATUS',
};
export { HmrRuntime };
/**
 * @record
 */
export function NgxsHmrSnapshot() { }
/**
 * @record
 * @template T
 */
export function NgxsHmrLifeCycle() { }
if (false) {
    /**
     * hmrNgxsStoreOnInit is called when the AppModule on init
     * @param {?} ctx - StateContext for the current app state from Store
     * @param {?} snapshot - previous state from Store after last hmr on destroy
     * @return {?}
     */
    NgxsHmrLifeCycle.prototype.hmrNgxsStoreOnInit = function (ctx, snapshot) { };
    /**
     * hmrNgxsStoreOnInit is called when the AppModule on destroy
     * @param {?} ctx - StateContext for the current app state from Store
     * @return {?}
     */
    NgxsHmrLifeCycle.prototype.hmrNgxsStoreBeforeOnDestroy = function (ctx) { };
}
/**
 * @record
 */
export function NgxsHmrOptions() { }
if (false) {
    /**
     * \@description
     * Clear logs after each refresh
     * (default: true)
     * @type {?|undefined}
     */
    NgxsHmrOptions.prototype.autoClearLogs;
    /**
     * \@description
     * Deferred time before loading the old state
     * (default: 100ms)
     * @type {?|undefined}
     */
    NgxsHmrOptions.prototype.deferTime;
}
/**
 * @record
 */
function WebpackHotModule() { }
if (false) {
    /** @type {?|undefined} */
    WebpackHotModule.prototype.hot;
}
