// angular-gojs/fesm5/angular-gojs.js
import { Component, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var GojsComponent = /** @class */ (function () {
    function GojsComponent() {
    }
    /**
     * @return {?}
     */
    GojsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    GojsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gojs-view',
                    template: "<p>\n  gojs\n</p>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    GojsComponent.ctorParameters = function () { return []; };
    return GojsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var GojsViewModule = /** @class */ (function () {
    function GojsViewModule() {
    }
    GojsViewModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [GojsComponent],
                    exports: [
                        GojsComponent
                    ]
                },] }
    ];
    return GojsViewModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { GojsViewModule, GojsComponent as ɵa };
